### PR TITLE
[opentitantool] SPI flash bootstrap updates 

### PR DESCRIPTION
**Pull request by alphan**
The first commit of this PR (authored by @cfrantz) adds the spi flash based bootstrap protocol while the second commit includes some minor changes to match the new bootstrap implementation.

### CODE REVIEW

File: sw/host/opentitanlib/src/bootstrap/eeprom.rs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    /// Performs the update protocol using the `transport` with the firmware `payload`.
+    fn update(
+        &self,
+        container: &Bootstrap,
+        transport: &TransportWrapper,
+        payload: &[u8],
+    ) -> Result<()> {
+        let spi = container.spi_params.create(transport)?;
+        let flash = SpiFlash::from_spi(&*spi)?;
+        flash.chip_erase(&*spi)?;
```
Comment (engdoreis) [inline sw/host/opentitanlib/src/bootstrap/eeprom.rs:48]: We are initializing the flash with a reference to the spi. So why do we still need to pass the spi in every call to the flash methods?

Maybe this comment is out of scope for this PR.

Comment (alphan) [reply sw/host/opentitanlib/src/bootstrap/eeprom.rs:48]: Just started looking at this part of opentitanlib. It looks like `SpiFlash` just holds the characteristics of the memory device and expects the actual IO to be used for communication passed in to its methods, i.e. dependency injection, but I'm not sure why it doesn't cache it at creation. @cfrantz might be able to clarify.

File: sw/host/opentitanlib/src/spiflash/flash.rs [hunk 2]
```diff
@@ -76,7 +76,7 @@ impl SpiFlash {
     pub const READ_SFDP: u8 = 0x5a;
     pub const NOP: u8 = 0x00;
     pub const RESET_ENABLE: u8 = 0x66;
-    pub const RESET: u8 = 0x99;
+    pub const RESET: u8 = 0xf0;
```
Comment (cfrantz) [inline sw/host/opentitanlib/src/spiflash/flash.rs:79]: I think the sequence RSTEN, RST (0x66, 0x99) is somewhat standard across flash parts.

- [MX66L2G45G](https://www.mxic.com.tw/Lists/Datasheet/Attachments/8697/MX66L2G45G,%203V,%202Gb,%20v1.0.pdf) Section 10-40, page 39.
- [W25Q128JV](https://www.winbond.com/resource-files/W25Q128JV%20RevI%2008232021%20Plus.pdf) Section 6.4, page 11.

If we're going to use a non-standard opcode and specify it in the SFDP, we should use the SFDP in the `chip_reset` function.

File: sw/host/opentitanlib/src/spiflash/flash.rs [hunk 3]
```diff
@@ -311,4 +319,10 @@ impl SpiFlash {
         }
         Ok(())
     }
+
+    /// Send the software reset sequence to the `spi` target.
+    pub fn chip_reset(spi: &dyn Target) -> Result<()> {
+        spi.run_transaction(&mut [Transfer::Write(&[Self::RESET])])?;
```
Comment (cfrantz) [inline sw/host/opentitanlib/src/spiflash/flash.rs:325]: My original error: this should be two transactions:
```
spi.run_transaction(&mut [Transfer::Write(&[Self::RESET_ENABLE])])?;
spi.run_transaction(&mut [Transfer::Write(&[Self::RESET])])?;
```

File: sw/host/opentitanlib/src/spiflash/flash.rs (hunk 2 above)
Comment (alphan) [reply sw/host/opentitanlib/src/spiflash/flash.rs:79]: The reason why we chose `0xf0` was that we wouldn't be able to enforce the 66/99 sequence anyway. 

Soft reset and rescue sequence support is defined in JESD216F.01, page 41 (bits 13:8 of the 16th DWORD).

We can also support both sequences or just the 66/99 if you prefer.

File: sw/host/opentitanlib/src/spiflash/flash.rs (hunk 3 above)
Comment (alphan) [reply sw/host/opentitanlib/src/spiflash/flash.rs:326]: Not really, I removed the RESET_ENABLE since we used a different opcode, reverted this change.

File: sw/host/opentitanlib/src/spiflash/flash.rs (hunk 2 above)
Comment (alphan) [reply sw/host/opentitanlib/src/spiflash/flash.rs]: Updated our RESET opcode to 0x99, ptal.
